### ISSUE TITLE
Directed network from neighborhood function

### ISSUE THREAD

**Issue by joanna-lada** [saezlab/pypath#99]
Dear Saez Lab,

I am an intern at the EBI; I am using omnipath to construct causal protein interaction networks, and in my work that it would be very useful to have a directed graph object as an output from 'neighborhood'. 
At present, the object produced is undirected, with the directionality information from different sources as a label attached to edges. I appreciate that there is ambiguity as to the directionality and sign of the causal interaction (activation/inhibition), so it would be great to have a stated method of deciding the directionality and sign, e.g. by majority consensus. At present, this is what I am doing edge-by-edge to construct a network.

A team member spoke to Dr Saez-Rodriguez today during his talk at EBI, was informed that this should be possible.

Thank you,

Joanna

**Comment by deeenes** [comment]
Hi Joanna,

To get a directed subnetwork let's say for BMP7 until 2nd neighbours:

```
from pypath import main, data_formats

pa = main.PyPath()
pa.init_network(data_formats.pathway)
sub_bmp7_2 = pa.get_directed(
    pa.graph.induced_subgraph(
        list(pa.gs_neighborhood('BMP7', order = 2).vs()
    )
)
``` 

This will be a directed `igraph` object, however the +/- signs are contained by the `Direction` objects.

```
import igraph

neighbours_1 = set(pa.gs_neighbors('BMP7').gs())

igraph.plot(
    sub_bmp7_2,
    layout = sub_bmp7_2.layout_fruchterman_reingold(
        repulserad = sub_bmp7_2.vcount() ** 2.1,
        maxiter = 1000,
        area = sub_bmp7_2.vcount() ** 2.3,
    ),
    vertex_color = [
        '#88CCAA'
            if v['label'] == 'BMP7' else
        '#DDDD77'
            if v['label'] in neighbours_1 else
        '#DDAA77'
        for v in sub_bmp7_2.vs()
    ],
    vertex_frame_width = .0,
    edge_color = [
        '#77112277'
            if e['dirs'].is_inhibition() else
        '#11777777'
            if e['dirs'].is_stimulation() else
        '#33333377'
        for e in sub_bmp7_2.es()
    ],
)
```

We have a method for majority vote directions and effects, but this won't make all edges unambiguous as still possible that the same number of resources support both directions. Also `get_directed` does not use this method (but actually we should add this option in the future!). Instead, for each edge you can call the method from the `Direction` object:

```
sub_bmp7_2.es['inhibition'] = [e['dirs'].is_inhibition() for e in sub_bmp7_2.es()]
sub_bmp7_2.es['stimulation'] = [e['dirs'].is_stimulation() for e in sub_bmp7_2.es()]

sub_bmp7_2.es['majority_inhibition'] = [
    e['dirs'].has_sign() and
    e['dirs'].majority_sign()[(
        sub_bmp7_2.vs[e.source]['name'],
        sub_bmp7_2.vs[e.target]['name'],
    )][1]
    for e in sub_bmp7_2.es()
]
sub_bmp7_2.es['majority_stimulation'] = [
    e['dirs'].has_sign() and
    e['dirs'].majority_sign()[(
        sub_bmp7_2.vs[e.source]['name'],
        sub_bmp7_2.vs[e.target]['name'],
    )][0]
    for e in sub_bmp7_2.es()
]

sub_bmp7_2.es['is_majority'] = [
    e['dirs'].majority_dir() == (
        sub_bmp7_2.vs[e.source]['name'],
        sub_bmp7_2.vs[e.target]['name']
    )
    for e in sub_bmp7_2.es()
]

# we have 10 mutual edges in this graph:
sum(sub_bmp7_2.is_mutual())
# we could exclude 8 of them by using the majority vote:
sub_bmp7_2.ecount() - sum(sub_bmp7_2.es['is_majority'])
```

Best,

Denes

**Comment by joanna-lada** [comment]
Hi Denes,
Thank you very much for this detailed response. After a bit more experimenting, I think I now get the distinction between neighborhood, 'affected_by/affects', 'stimulated/inhibits' etc with respect to the annotations available.

I think I have also spotted an error in src/pypath/main.py line 8559 onward:

   ` def up_neighborhood(self, **uniprot**, order=1, mode='ALL'):
        """
        """

        if type(uniprots) in common.simpleTypes:`

In bold should be 'uniprots' as using

'pa.up_neighborhoods' returns the error

 if type(uniprots) in common.simpleTypes:
UnboundLocalError: local variable 'uniprots' referenced before assignment


Thanks again,

Joanna

**Comment by deeenes** [comment]
Great, I also realized it would be convenient to have some more methods like this, we will create them later. Thanks for finding the typo, fixed in f0e330f

Best,

Denes
